### DISCUSSION TITLE
Update `--sourcemap` to enable different source map generation modes

### DISCUSSION THREAD

**Discussion by ghost** [remix-run/remix#7395] [Proposals]
**What**
[`remix build` command only supports one way to generate source maps](https://github.com/remix-run/remix/blob/main/packages/remix-dev/compiler/options.ts#L5), however [ESBuild supports four different modes of source map generation](https://esbuild.github.io/api/#sourcemap)

I would like to see more flexibility here for developers, specifically I'd like to see Remix support the `external` option.

> **external**
> 
> This mode means the source map is generated into a separate .js.map output file alongside the .js output file, but unlike linked mode the .js output file does not contain a //# sourceMappingURL= comment. Use external source map mode like this:
> ```bash
> esbuild app.ts --sourcemap=external --outfile=out.js
> ```
source: https://esbuild.github.io/api/#sourcemap

**Why**

I want to improve the experience of generating and make using source maps in production better.

Currently, `remix build --sourcemap` is the only way to generate source maps. The problem with this approach is that it leaves a bunch of `//# sourceMappingURL=...` comments in production files. These comments produce a bunch of 404 errors in production when the browser console is open.

![image](https://github.com/remix-run/remix/assets/94476616/a3f7d388-dd0b-4fd6-8fbb-013d3d453786)

If you're uploading your source maps to an external provider or hosting location (like [Sentry.io](https://sentry.io/)) then these `//# sourceMappingURL=...` comments in source files are not needed and can be removed.

**Comment by ushironoko** [comment]
I am running into the same problem. remix deprecates hosting the sourcemap in production, but if the `//# sourceMappingURL=...` is present, even if the chunk's map file is removed as instructed, a request to the map file is still generated. request to the map file will be generated even if the chunk's map file is deleted as described in the procedure.
https://docs.sentry.io/platforms/javascript/guides/remix/sourcemaps/#3-remove-remix-source-maps

You can disable the reference to `//# sourceMappingURL=...`, but it may be difficult to do so easily since it is an org-wide setting.

**Comment by wasanx25** [comment]
I have same issue. We need option of remove source map comment too like this https://github.com/remix-run/remix/discussions/7352.
